metadata.AppendToOutgoingContext may cause the metadata in parent context to be modified

### What version of gRPC are you using?

    master branch:  fa28bef9392c6c3e28e75389d8be8a6797561f57

### What version of Go are you using (`go version`)?

    go version go1.10 darwin/amd64

### What operating system (Linux, Windows, …) and version?

    macOS High Sierra 
    Version 10.13.3

### What did you do?

```go
package main

import (
	"context"
	"fmt"
	"reflect"
	"testing"

	"google.golang.org/grpc/metadata"
)

func TestAppendToOutgoingContext(t *testing.T) {
	var ctx = context.Background()

	var ctx1 context.Context
	var ctx2 context.Context

	for i := 0; i < 100; i = i + 2 {
		ctx1 = metadata.AppendToOutgoingContext(ctx, fmt.Sprintf("k1-%d", i), fmt.Sprintf("v1-%d", i+1))
		ctx2 = metadata.AppendToOutgoingContext(ctx, fmt.Sprintf("k2-%d", i), fmt.Sprintf("v2-%d", i+1))

		md1, _ := metadata.FromOutgoingContext(ctx1)
		md2, _ := metadata.FromOutgoingContext(ctx2)

		if reflect.DeepEqual(md1, md2) {
			panic(fmt.Sprintf("md1:(%v) and md2:(%v) should not be equal", md1, md2))
		}

		ctx = ctx1
	}
}
```

This test case failed with:

```shell
--- FAIL: TestAppendToOutgoingContext (0.00s)
panic: md1:(map[k1-0:[v1-1] k1-2:[v1-3] k1-4:[v1-5] k2-6:[v2-7]]) and md2:(map[k1-0:[v1-1] k1-2:[v1-3] k1-4:[v1-5] k2-6:[v2-7]]) should not be equal [recovered]

goroutine 5 [running]:
testing.tRunner.func1(0xc4200a20f0)
	GOROOT/src/testing/testing.go:711 +0x2d2
panic(0x110c5c0, 0xc42000e920)
	GOROOT/src/runtime/panic.go:491 +0x283
tests/grpc-go.TestAppendToOutgoingContext(0xc4200a20f0)
	tests/grpc-go/metadata_test.go:26 +0x691
testing.tRunner(0xc4200a20f0, 0x1144398)
	GOROOT/src/testing/testing.go:746 +0xd0
created by testing.(*T).Run
	GOROOT/src/testing/testing.go:789 +0x2de
```

### What did you expect to see?

    Unit test TestAppendToOutgoingContext pass.

### What did you see instead?

    https://github.com/grpc/grpc-go/blob/fa28bef9392c6c3e28e75389d8be8a6797561f57/metadata/metadata.go#L134

```go
rawMD{md: md.md, added: append(md.added, kv)}
```

This may cause md.added in parent ctx to be modified.

--- a/metadata/metadata.go
+++ b/metadata/metadata.go
@@ -131,7 +131,11 @@ func AppendToOutgoingContext(ctx context.Context, kv ...string) context.Context 
 		panic(fmt.Sprintf("metadata: AppendToOutgoingContext got an odd number of input pairs for metadata: %d", len(kv)))
 	}
 	md, _ := ctx.Value(mdOutgoingKey{}).(rawMD)
-	return context.WithValue(ctx, mdOutgoingKey{}, rawMD{md: md.md, added: append(md.added, kv)})
+	added := make([][]string, len(md.added)+1)
+	copy(added, md.added)
+	added[len(added)-1] = make([]string, len(kv))
+	copy(added[len(added)-1], kv)
+	return context.WithValue(ctx, mdOutgoingKey{}, rawMD{md: md.md, added: added})
 }
 
 // FromIncomingContext returns the incoming metadata in ctx if it exists.  The
@@ -159,7 +163,7 @@ func FromOutgoingContextRaw(ctx context.Context) (MD, [][]string, bool) {
 
 // FromOutgoingContext returns the outgoing metadata in ctx if it exists.  The
 // returned MD should not be modified. Writing to it may cause races.
-// Modification should be made to the copies of the returned MD.
+// Modification should be made to copies of the returned MD.
 func FromOutgoingContext(ctx context.Context) (MD, bool) {
 	raw, ok := ctx.Value(mdOutgoingKey{}).(rawMD)
 	if !ok {

--- a/metadata/metadata_test.go
+++ b/metadata/metadata_test.go
@@ -134,30 +134,25 @@ func TestAppendToOutgoingContext_FromKVSlice(t *testing.T) {
 
 // Old/slow approach to adding metadata to context
 func Benchmark_AddingMetadata_ContextManipulationApproach(b *testing.B) {
-	for _, num := range []int{1, 10, 100} {
-		b.Run(strconv.Itoa(num), func(b *testing.B) {
-			for n := 0; n < b.N; n++ {
-				ctx := context.Background()
-				for i := 0; i < num; i++ {
-					md, _ := FromOutgoingContext(ctx)
-					NewOutgoingContext(ctx, Join(Pairs("k1", "v1", "k2", "v2"), md))
-				}
-			}
-		})
+	// TODO: Add in N=1-100 tests once Go1.6 support is removed.
+	const num = 10
+	for n := 0; n < b.N; n++ {
+		ctx := context.Background()
+		for i := 0; i < num; i++ {
+			md, _ := FromOutgoingContext(ctx)
+			NewOutgoingContext(ctx, Join(Pairs("k1", "v1", "k2", "v2"), md))
+		}
 	}
 }
 
 // Newer/faster approach to adding metadata to context
 func BenchmarkAppendToOutgoingContext(b *testing.B) {
-	for _, num := range []int{1, 10, 100} {
-		b.Run(strconv.Itoa(num), func(b *testing.B) {
-			for n := 0; n < b.N; n++ {
-				ctx := context.Background()
-				for i := 0; i < num; i++ {
-					ctx = AppendToOutgoingContext(ctx, "k1", "v1", "k2", "v2")
-				}
-			}
-		})
+	const num = 10
+	for n := 0; n < b.N; n++ {
+		ctx := context.Background()
+		for i := 0; i < num; i++ {
+			ctx = AppendToOutgoingContext(ctx, "k1", "v1", "k2", "v2")
+		}
 	}
 }
 

--- a/metadata/metadata_test.go
+++ b/metadata/metadata_test.go
@@ -20,6 +20,7 @@ package metadata
 
 import (
 	"reflect"
+	"strconv"
 	"testing"
 
 	"golang.org/x/net/context"
@@ -98,19 +99,65 @@ func TestAppendToOutgoingContext(t *testing.T) {
 	}
 }
 
+func TestAppendToOutgoingContext_Repeated(t *testing.T) {
+	ctx := context.Background()
+
+	for i := 0; i < 100; i = i + 2 {
+		ctx1 := AppendToOutgoingContext(ctx, "k", strconv.Itoa(i))
+		ctx2 := AppendToOutgoingContext(ctx, "k", strconv.Itoa(i+1))
+
+		md1, _ := FromOutgoingContext(ctx1)
+		md2, _ := FromOutgoingContext(ctx2)
+
+		if reflect.DeepEqual(md1, md2) {
+			t.Fatalf("md1, md2 = %v, %v; should not be equal", md1, md2)
+		}
+
+		ctx = ctx1
+	}
+}
+
+func TestAppendToOutgoingContext_FromKVSlice(t *testing.T) {
+	const k, v = "a", "b"
+	kv := []string{k, v}
+	ctx := AppendToOutgoingContext(context.Background(), kv...)
+	md, _ := FromOutgoingContext(ctx)
+	if md[k][0] != v {
+		t.Fatalf("md[%q] = %q; want %q", k, md[k], v)
+	}
+	kv[1] = "xxx"
+	md, _ = FromOutgoingContext(ctx)
+	if md[k][0] != v {
+		t.Fatalf("md[%q] = %q; want %q", k, md[k], v)
+	}
+}
+
 // Old/slow approach to adding metadata to context
 func Benchmark_AddingMetadata_ContextManipulationApproach(b *testing.B) {
-	for n := 0; n < b.N; n++ {
-		ctx := context.Background()
-		md, _ := FromOutgoingContext(ctx)
-		NewOutgoingContext(ctx, Join(Pairs("k1", "v1", "k2", "v2"), md))
+	for _, num := range []int{1, 10, 100} {
+		b.Run(strconv.Itoa(num), func(b *testing.B) {
+			for n := 0; n < b.N; n++ {
+				ctx := context.Background()
+				for i := 0; i < num; i++ {
+					md, _ := FromOutgoingContext(ctx)
+					NewOutgoingContext(ctx, Join(Pairs("k1", "v1", "k2", "v2"), md))
+				}
+			}
+		})
 	}
 }
 
 // Newer/faster approach to adding metadata to context
 func BenchmarkAppendToOutgoingContext(b *testing.B) {
-	for n := 0; n < b.N; n++ {
-		AppendToOutgoingContext(context.Background(), "k1", "v1", "k2", "v2")
+	for _, num := range []int{1, 10, 100} {
+		b.Run(strconv.Itoa(num), func(b *testing.B) {
+			for n := 0; n < b.N; n++ {
+				ctx := context.Background()
+				for i := 0; i < num; i++ {
+					ctx = AppendToOutgoingContext(ctx, "k1", "v1", "k2", "v2")
+				}
+			}
+		})
 	}
 }
 

